Bug #80545: Converted some type warnings to type errors in the "bcmath" extension.

jensdenies opened the pull request:
Hi,

This my first attempt to solve one of the bugs from the PHP bug tracker (#80545). I targeted this to the PHP 8.0 branch because the warning promotion didn't take place yet in PHP 7.4 (please correct me if I'm wrong).

Bccomp had the same issue, so I changed it there too.

If there's anything I did wrong or needs to change, please let me know! :)

cmb69 commented:
That is analogous to what GMP does, but in my opinion, it would be beneficial to have a dedicated function to verify the validity of a number given as string. Also, I think a ValueError would be more appropriate than a TypeError.

jensdenies commented:
@cmb69 I will change it to a value error, that indeed looks more appropriate.

There's a function called php_str2num in the bcmath.c file, or would you like to have a globally accessible function?

cmb69 commented:
> I will change it to a value error, that indeed looks more appropriate.

I'd rather be consistent with GMP. @Girgias, since you [promoted to TypeError](https://github.com/php/php-src/commit/e208cb23c6bda00737a9bfd227b82beaba350fa7), what do you think? /cc @nikic 

> There's a function called php_str2num in the bcmath.c file, or would you like to have a globally accessible function?

Yes, this was about a function which could be called from userland. And that function should just return true or false, similar to `is_numeric()`. Anyhow, a new function should likely have its own PR.

Girgias commented:
I think a ValueError would be better, not exactly sure why I did a TypeError for GMP, but I need to possibly tweak some things with GMP anyway.

Would it be possible to use the version with the argument number (IIRC ``zend_argument_value_error()``) ? As this is usually better IMHO.

cmb69 commented:
> Would it be possible to use the version with the argument number (IIRC zend_argument_value_error()) ?

That would require to pass the arg num to `php_str2num()`, but that shouldn't be an issue.

jensdenies commented:
@cmb69 @Girgias I changed the `TypeError` to a `ValueError` and I've used the `zend_argument_value_error ` variant.

jensdenies commented:
Not sure why the CI failed.

cmb69 commented:
I think that CI failure is unrelated to this PR.

nikic commented on ext/bcmath/bcmath.c:
```diff
@@ -174,8 +174,8 @@ PHP_FUNCTION(bcadd)
 	bc_init_num(&result);
-	php_str2num(&first, ZSTR_VAL(left));
-	php_str2num(&second, ZSTR_VAL(right));
+	php_str2num(&first, ZSTR_VAL(left), 1);
```
php_str2num should return whether conversion was successful. On failure, we should do an early `RETURN_THROWS();` here. Otherwise the rest of the function will continue to execute. If nothing else, this will result in a double exception if both arguments are not well-formed.

jensdenies replied:
@nikic Do I need to free `first` and `second` before returning?

cmb69 replied:
Actually, all three numbers should be `bc_num_free()`d on failure. Might make sense to introduce some `goto finish;` (or similar) here. Still the function must only `RETURN_THROWS()` on failure, so the failure would need to get stored in a local variable for this to work.

jensdenies commented:
@nikic Thanks, I changed the return type from `void` to `zend_result` and moved the value error to the functions itself. I also added the early returns for the other functions in this file. Also noticed some empty returns at the end of some functions, I removed those as they seem unnecessary. The tests still pass locally.

nikic commented on ext/bcmath/bcmath.c:
```diff
@@ -174,15 +177,23 @@ PHP_FUNCTION(bcadd)
+
+	if (php_str2num(&first, ZSTR_VAL(left)) == FAILURE) {
+		zend_argument_value_error(1, "bcmath function argument is not well-formed");
+		RETURN_THROWS();
```
Replace `RETURN_THROWS()` with `goto cleanup;` and then add `cleanup:` before the three `bc_free_num` call at the end of the function. Similar for the other cases.

You should also be able to add a test that currently leaks, e.g. using `bcadd("1", "a");` i.e. where the first parameter is correct but the second one isn't.

jensdenies replied:
@nikic Thanks, I added the `goto`'s and added a test. I'm wondering why a memory leak only occurs when the second parameter is invalid and not when the first parameter is invalid?

nikic replied:
Presumably bc_init_num by itself doesn't allocate anything, so we need to the first number to be initialized to something non-trivial in order to leak.

cmb69 replied:
`bc_init_num` copies (i.e. increases the ref count of) a module global. I think there is a theoretical leak on GSHUTDOWN.

nikic commented on ext/bcmath/bcmath.c:
```diff
@@ -174,15 +177,26 @@ PHP_FUNCTION(bcadd)
-	php_str2num(&second, ZSTR_VAL(right));
+
+	if (php_str2num(&first, ZSTR_VAL(left)) == FAILURE) {
+		zend_argument_value_error(1, "bcmath function argument is not well-formed");
```
```suggestion
		zend_argument_value_error(1, "is not well-formed");
```
This will make the error message:
> bcadd(): Argument #1 ($num1) is not well-formed

jensdenies replied:
Thanks @nikic, fixed.

nikic reviewed (commented):
This looks good to me, I'd only suggest to adjust the message to fix the "argument_value_error" format better.

jensdenies commented:
@nikic I've pointed this PR to the PHP-8.0 branch, is that correct?

nikic commented:
I merged this as https://github.com/php/php-src/commit/94a151a018d150ff6470fc35bc509f1bd8e3956a a while ago, but messed up the auto-close tag.
